project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/services/GitWrapperService.kt

```diff
@@ -0,0 +1,127 @@
+package org.plan.research.minimization.plugin.services
+
+import  org.plan.research.minimization.plugin.model.IJDDContext
+
+import com.intellij.openapi.application.readAction
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.project.isProjectOrWorkspaceFile
+import com.intellij.openapi.util.io.FileUtil
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.toNioPathOrNull
+import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.api.ResetCommand
+import org.eclipse.jgit.api.errors.NoHeadException
+
+import java.io.File
+import java.util.*
+
+import kotlin.io.path.pathString
+import kotlin.io.path.relativeTo
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+/**
+ * Service responsible for git operations within ProjectGitSnapshotManager.
+ */
+@Service(Service.Level.APP)
+class GitWrapperService {
+    private val importantFiles = setOf("modules.xml", "misc.xml", "libraries")
+
+    suspend fun commitChanges(context: IJDDContext): IJDDContext = context.apply {
+        commit(this.projectDir)
+        projectDir.refresh(false, true)
+    }
+
+    suspend fun resetChanges(context: IJDDContext) {
+        withContext(Dispatchers.IO) {
+            // git reset --hard
+            Git.open(context.projectDir.toNioPath().toFile()).apply {
+                reset().setMode(ResetCommand.ResetType.HARD).call()
+                clean().setCleanDirectories(true).call()
+                close()
+            }
+        }
+    }
+
+    private suspend fun commit(projectDir: VirtualFile) = withContext(Dispatchers.IO) {
+        Git.open(projectDir.toNioPath().toFile()).let {
+            projectDir.gitAdd(it) { file ->
+                isImportant(file, projectDir)
+            }
+            it.commit().setMessage(UUID.randomUUID().toString()).setAllowEmpty(true)
+                .call()
+            it.close()
+        }
+    }
+
+    private fun isImportant(file: VirtualFile, root: VirtualFile): Boolean {
+        val path = file.toNioPath().relativeTo(root.toNioPath())
+        if (isProjectOrWorkspaceFile(file) && file.name != Project.DIRECTORY_STORE_FOLDER) {
+            val pathString = path.pathString
+            return importantFiles.any { it in pathString }
+        }
+        return true
+    }
+
+    fun gitInitOrOpen(VFProjectDir: VirtualFile): Git {
+        val projectDir: File = VFProjectDir.toNioPath().toFile()
+        if (projectDir.resolve(".git").exists()) {
+            return Git.open(projectDir)
+        }
+        return Git.init()
+            .setDirectory(projectDir)
+            .call()
+            .apply {
+                commit().setMessage("init commit").call()
+            }
+    }
+
+    private fun isCommitListEmpty(git: Git): Boolean = try {
+        git.log().call()  // This will throw NoHeadException if there are no commits
```

Comment:
For the getting I propose to set `setMaxCount(1)`, since `git log` could work very slowly on the project with big history. E.g. `IJ Ultimate`:
```zsh
❯ /usr/bin/time git --no-pager log -n 1 > /dev/null
        0.04 real         0.00 user         0.03 sys

ultimate on master 
❯ /usr/bin/time git --no-pager log > /dev/null 
       16.48 real        12.69 user         0.97 sys
```